### ISSUE TITLE
How do I easily extend this to add my own export format?

### ISSUE THREAD

**Issue by tacticiankerala** [angra007/AliensOnEarth#1]
I could not find how you have implemented the Part B of the problem,

Part B: 
Make the application extensible so that it is easy to “plug in” a new format to export to, so that  the application automatically has the capability to export to the new format without making any changes to the application code.
Example - Google does not make changes to Chrome each time a new plugin for Chrome is published. The plugin effectively provides new capability without changing any code in Chrome.

**Comment by angra007** [comment]
Hello Sreenath ,

Thank you for the review.

In this assignment ExportAlienData.h/.m is the super class for all the
export related work. All the formats will inherit from this class . The
base classes provide export logic where as super class provide data to them.

However, I have a better logic (using protocols). I will update the code
and let you know.

Thank You.

Regards,
Ankit Angra
8880213122

**Comment by tacticiankerala** [comment]
:+1: Let me know :)

**Comment by angra007** [comment]
Hello Sreenath,

I have updated the code and mailed to Maria (with explanation of current
and previous version).

Please provide me feedback/suggestion for what I could have done
differently.

Thank You.

Regards,
Ankit Angra
+91-8880213122

**Comment by tacticiankerala** [comment]
One important suggestion I have is to add a proper README. For example it is not clear that how can i add a new format without making a lot of modifications.

Also, I am not able to understand why some classes have the word "object" in its name and some have "class" in its name. Any particular reason?

**Comment by angra007** [comment]
The pattern I have tried to follow is that every format type has its own
class. This is because I didn't wanted to put every this in one class as it
would be very difficult to maintain. The only way we could maintain this
kind of code if by adding "if else" conditions which would have made code
very complex. Referring to the example of Google Chrome Plugin. Chrome
would have some kind of mapping i.e when the user clicks on a particular
plugin which code to call (I have never worked with them. It is just a
guess).  If I want to extend this code in the future I have to do similar
kind of mapping which Google Chrome might do under the hood.

Moreover, in iOS a common way to add any new thing in your code is to use
define a protocol and implement its methods. For example, UITableView in iOS
has defined a protocol  which has several methods in it. If we want to
implement a new table view every time we have to conform to that protocol
and implement those methods (as I have done in this assignment). We can
however modify, its logic but basic methods should be implemented. Similarly
these are many build-in class which follow the similar design pattern.

Referring to your second point. All the models are defined as objects. This
is the naming conventions I was told to follow and this has been used in
multiple of our project..  I think it is just style that objective-c follows
but, I am not sure about it. For example, if we define a custom cell for
table view , we would generally end the name of the class by CELL (
TravelExpenseCell). There is no hard and past rule for it, but I think it
is just done to clearly differentiate what kind of class it is.
Thank You,

Regards,
Ankit Angra

**Comment by tacticiankerala** [comment]
I totally agree that 'TravelExpenseCell' is a meaningful name. But what  about 'TravelExpenseCellClass' and 'TravelExpenseCellObject'? what do they really convey? And the following are also widely used conventions,

- Use nouns for class names and property names
- Use verbs for method names to describe action

Can you re-look at your code and see if there is anything to improve on this aspect?

Coming to the other part, to add a google chrome plugin, we do not modify google chrome code right? Can you say that the same thing will be true for this application?

**Comment by angra007** [comment]
I'll look into the naming conventions and try to improve  them.

For Google Chrome, NO same is not true for my application. This is because
I need some way my class which takes the input , to take to class which
will manage the export. Thus every-time I add a new format type I need to
add a new case in the switch statement and add that type in enum and
exportmanager willl handle the rest. In case of a plugin we do not do this
mapping (I guess).  I have no idea about Plugins but I can give you a
similar example in Objective-C.
When ever we want to use a button we drag it from Interface Builder and
hook it into the code. The internal mapping is done by iOS. Thus
Objective-C knows which method to call when hit a button. Objective-C is
able to do it because UIButton in interface builder.I suppose in Google
Chrome there might be similar implementation. We would add a button and
that button will know which method to call.


Regards,
Ankit Angra

**Comment by tacticiankerala** [comment]
Great!!

 Can you also look into [SOLID](http://www.codeproject.com/Articles/60845/The-S-O-L-I-D-Object-Oriented-Programming-OOP-Prin) principles and try to apply them in your code.

**Comment by angra007** [comment]
Sure. I'll check if I can improve the design.

Thank You.

Regards,
Ankit Angra

**Comment by angra007** [comment]
Hi Sreenath,


Greetings for the day.

This is just to inform you that I have found a way to do it. But
unfortunately, I occupied with some of the deliverables. I'll complete it
this weekend and let you know.

Thank You,

Regards,
Ankit Angra
+91-8880213122

**Comment by angra007** [comment]
Hello Sreenath,


Greetings for the day.

I have tried to implement objective-c reflection. Please refer to link:
https://github.com/angra007/AlienOnEarth

Thank You.

Regards,
Ankit Angra
+91-8880213122

**Comment by tacticiankerala** [comment]
Sorry for the delay. I was out of town for couple of days. 

Great to hear that you were able to make progress :+1: 

I will go through the code and get back to you as soon as possible.
